File: C Language/queue_menu.c
/*
 * queue_menu.c
 * This program manages a menu of common queue operations.
 */

#include "queue.h"   // for Queue and its functions

#include <stdio.h>   // for printf, scanf, getchar, puts
#include <stdlib.h>  // for malloc, free
#include <stddef.h>  // for NULL

// Declares functions using prototypes.
void show_menu(void);
int read_selection(void);
void process_selection(int sel, Queue *que);
void process_enqueue(Queue *que);
void process_dequeue(Queue *que);
void process_view(const Queue *que);
void show_element(void *elem);
void destroy_element(void *elem);

// Starts the execution of the program.
int main(void) {
    Queue *que = create_queue();
    int sel = 0;

    do {
        show_menu();
        sel = read_selection();
        process_selection(sel, que);
    } while (sel != 0);

    iterate_queue(que, &destroy_element);
    destroy_queue(que);
    que = NULL;
}

// Shows the menu options.
void show_menu(void) {
    puts("\nQueue Menu\n");
    puts("1. Enqueue an element");
    puts("2. Dequeue an element");
    puts("3. View the contents");
    puts("0. Exit");
}

// Returns the menu option selected by the user.
int read_selection(void) {
    int sel;
    printf("Enter your selection: ");
    scanf("%d", &sel);
    while (getchar() != '\n');  // empty input buffer
    return sel;
}

// Processes a menu selection and updates a Queue.
void process_selection(int sel, Queue *que) {
    switch (sel) {
        case 1: process_enqueue(que); break;
        case 2: process_dequeue(que); break;
        case 3: process_view(que); break;
        case 0: puts("Thanks for using this program!"); break;
        default: puts("Error! Selection must be between 0 and 3.");
    }
}

// Process the push element menu option.
void process_enqueue(Queue *que) {
    char *elem = malloc(sizeof(char));
    printf("Enter a character: ");
    scanf("%c", elem);
    enqueue(que, elem);
    printf("%c has been enqueued.\n", *elem);
}

// Process the pop element menu option.
void process_dequeue(Queue *que) {
    char *elem = dequeue(que);
    if (elem == NULL)
        puts("Could not dequeue. Queue is empty.");
    else {
        printf("%c has been dequeued.\n", *elem);
        free(elem);
    }
}

// Process the view contents menu option.
void process_view(const Queue *que) {
    if (is_empty_queue(que))
        puts("Cannot show.  Queue is empty.");
    else {
        printf("These are the contents of the queue: ");
        iterate_queue(que, &show_element);
        printf("\n");
    }
}

// Displays an element.
void show_element(void *elem) {
    printf("%c ", *(char *)elem);
}

// Deallocates the memory of an element.
void destroy_element(void *elem) {
    free(elem);
}

File: Prolog Language/queue_menu.pl
/*
 * queue_menu.pl
 * This program manages a menu of common queue operations.
 */

:-use_module(queue_lst).
:-initialization(main).

%% main
%  Serves as an entry point of the program.
%  Creates an empty queue and manages the menu.
main :-
    empty_queue(Queue),
    menu(Queue).

%% menu(+Queue)
%  Displays the menu, reads the selection and process this selection.
menu(Queue) :-
    write('\n\tQueue Menu\n\n'),
    write('1. Enqueue an element\n'),
    write('2. Dequeue an element\n'),
    write('3. View the contents\n'),
    write('4. Quit\n'),
    write('Enter your selection (1-4, ending with a full stop): '),
    read(Selection), nl,
    selection(Selection, Queue).

%% selection(+Selection, +Queue)
%  Processes the menu selection with the current queue.
selection(1, Queue) :-
    write('Enter an element (end with full stop): '),
    read(Elem),
    enqueue(Queue, Elem, NewQueue),
    write(Elem), write(' has been enqueued.\n'),
    menu(NewQueue).

selection(2, Queue) :-
    catch(
        (
            dequeue(Queue, Elem, NewQueue),
            write(Elem), write(' has been dequeued.\n')
        ),
        empty_exception(Message),
        (
            write('Cannot dequeue: '), write(Message), nl,
            NewQueue = Queue
        )
    ),
    menu(NewQueue).

selection(3, Queue) :-
    empty_queue(Queue),
    write('Cannot show: queue is empty.\n'),
    menu(Queue).

selection(3, Queue) :-
    \+ empty_queue(Queue),
    write('These are the contents of the queue: '),
    iterate_queue(Queue, write_element), nl,
    menu(Queue).

selection(4, _) :-
    write('Thanks for using this program.\n').

selection(Selection, Queue) :-
    (
        Selection < 1
        ;
        Selection > 4
    ),
    write('Error! Selection must be between 1 and 4.\n'),
    menu(Queue).

%% write_element(+Elem)
%  Displays an element followed by a space.
write_element(Elem) :-
    write(Elem), write(' ').